### ISSUE TITLE
Possibly-undefined: allow assertions on the lengths of sequences

### ISSUE THREAD

**Issue by AlexWaygood** [python/mypy#14830]
The new (disabled-by-default) `possibly-undefined` error code is great, and I'm using it in several of my projects. There are many hits if you enable it for typechecking mypy's own code, however. An example is this function, where mypy (correctly) points out that `arg` may be undefined on line 526 (`arg_types` might be an empty list).

https://github.com/python/mypy/blob/456dcbdd6a2f9b36249e0e7d31e521f5678a8235/mypy/suggestions.py#L509-L527

It would be nice if we could use assertions based on the truthiness of a sequence to make this error go away. E.g., it would be nice if we could silence the error here with this diff:

```diff
--- a/mypy/suggestions.py
+++ b/mypy/suggestions.py
@@ -512,6 +512,7 @@ class SuggestionEngine:
         arg_names: list[list[str | None]],
         arg_types: list[list[Type]],
     ) -> str:
+        assert arg_types
         args: list[str] = []
```

Mypy doesn't currently understand this idiom, however. @ilinum, how hard might this kind of thing be to implement?

**Comment by JelleZijlstra** [comment]
For your example, wouldn't we instead have to assert that `zip(arg_kinds[i], arg_names[i], arg_types[i])` is nonempty? `args` is defined whether or not `arg_types` is empty.

The feature could still be useful in other cases, but not sure how common the need would be. When I implemented this feature in pyanalyze I put in some features that can detect that certain iterables are always nonempty, e.g., certain calls to `range()`.

**Comment by AlexWaygood** [comment]
> For your example, wouldn't we instead have to assert that `zip(arg_kinds[i], arg_names[i], arg_types[i])` is nonempty? `args` is defined whether or not `arg_types` is empty.

Good point, that's a terrible example I gave! But you get my drift.

**Comment by cdce8p** [comment]
> IIRC, someone (@cdce8p?) was talking about container types and how one could declare it as non-empty (supporting `NonEmpty[List[int]]` or something like that).

Yeah, there is an open issue for it #9745. I think something like `NonEmpty` would really help with that issue. It's not just limited to mypy btw as pyright as a similar check with `possibly unbound` that would benefit from it.

The tricky part as you mentioned would be to track changes but that should be possible as well, at the very least for a majority of cases. We could even be quite strict about it: Only assume `NonEmpty` if we know it for sure, as soon as the length could decrease, e.g. with `pop()`, it's no longer a given.

Some more details: https://github.com/python/mypy/issues/9745#issuecomment-1356353819
